internal/sysenc/marshal.go

```diff
@@ -0,0 +1,151 @@
+package sysenc
+
+import (
+	"bytes"
+	"encoding"
+	"encoding/binary"
+	"errors"
+	"fmt"
+	"reflect"
+	"sync"
+	"unsafe"
+
+	"github.com/cilium/ebpf/internal"
+)
+
+var bytesBufferPool = sync.Pool{
+	New: func() interface{} {
+		return new(bytes.Buffer)
+	},
+}
+
+// Marshal turns data into a byte slice using the system's native endianness.
+//
+// Returns an error if the data can't be turned into a byte slice according to
+// the behaviour of [binary.Write].
+func Marshal(data any, size int) (Buffer, error) {
+	if data == nil {
+		return Buffer{}, errors.New("can't marshal a nil value")
+	}
+
+	layout := dataLayout{1, size, size}
+
+	var buf []byte
+	var err error
+	switch value := data.(type) {
+	case encoding.BinaryMarshaler:
+		buf, err = value.MarshalBinary()
+	case string:
+		// Replace with unsafe.Slice(unsafe.StringData()) once we target go 1.20.
+		buf = []byte(value)
+	case []byte:
+		buf = value
+	case int16:
+		buf = internal.NativeEndian.AppendUint16(make([]byte, 0, 2), uint16(value))
+	case uint16:
+		buf = internal.NativeEndian.AppendUint16(make([]byte, 0, 2), value)
+	case int32:
+		buf = internal.NativeEndian.AppendUint32(make([]byte, 0, 4), uint32(value))
+	case uint32:
+		buf = internal.NativeEndian.AppendUint32(make([]byte, 0, 4), value)
+	case int64:
+		buf = internal.NativeEndian.AppendUint64(make([]byte, 0, 8), uint64(value))
+	case uint64:
+		buf = internal.NativeEndian.AppendUint64(make([]byte, 0, 8), value)
+	default:
+		if buf := unsafeBackingMemory(data, layout); buf != nil {
+			return newBuffer(buf, layout), nil
+		}
+
+		wr := bytesBufferPool.Get().(*bytes.Buffer)
+		defer bytesBufferPool.Put(wr)
+
+		// Reinitialize the Buffer with a new backing slice since it is returned to
+		// the caller by wr.Bytes() below. Pooling is faster despite calling
+		// NewBuffer. The pooled alloc is still reused, it only needs to be zeroed.
+		*wr = *bytes.NewBuffer(make([]byte, 0, size))
+
+		err = binary.Write(wr, internal.NativeEndian, value)
```

Comment:
Please forgive me my ignorance, but won't it be much easier to verify compatibility by round-tripping a buffer of `Sizeof(T)` `FF`s via `binary.Read` followed by `binary.Write` and checking that we got the same value back?

Reply:
Well, comparing `Sizeof(T)` with the size of blob produced/consumed by `binary.Write()` / `binary.Read()` will tell whether `T` has any compiler-inserted padding.

It still remains to recursively check `T` to see if there are any struct fields that are treated as padding by `encoding/binary`, but tracking offsets/sizes won't be necessary.

Reply:
You mean comparing `reflect.Type.Size()` with `binary.Size`? True, that would be easier, but it doesn't allow detecting the case where a struct has implicit trailing padding. The current code allows zero copy from something like this:

```go
struct {
    A uint32
    B uint16
    // two bytes of implicit padding
}
```

You're right that the `dataLayout` machinery is kind of involved. The reason it's this way is because I was looking at doing zero-copy marshaling for per-CPU maps as well. It turned out too big of a change at once though, but it's something I'd like to revisit at some point.

Reply:
> The current code allows zero copy from something like this:

I'm curious when this is beneficial. My assumption is that this patch doesn't change any user-visible behaviours, apart from making things faster.

`marshalBytes` checks that resulting buf has the size as expected. With trailing implicit padding it won't be the case, unless C structure is packed. Indeed, `binary.Write` will produce 6 bytes while 8 are expected. Packed structs aren't that abundant to make it worthwhile, are they?

`unmarshalBytes` on the other hand is ok if fewer bytes are consumed. Is that the case you are optimising for?

> I was looking at doing zero-copy marshaling for per-CPU maps as well

Trying to wrap my head around it, does it mean reading into correctly sized slice from bpf syscall directly? Will it work for updates as well? Presumably currently it is ok to pass a shorter slice.

Reply:
> My assumption is that this patch doesn't change any user-visible behaviours, apart from making things faster.

That's the intention!

> Packed structs aren't that abundant to make it worthwhile, are they?

Okay, I think you're saying the following: given a C structure that has implicit padding, we'd declare a map that has a key / value length which includes the padding. On the Go side we'd be looking for a length of 8 instead of 6. So the fast past wouldn't apply, in fact we'd bail out of marshaling since the Go struct doesn't have the correct length.

This is true, I hadn't thought of that. My angle is that the fast path should apply for a struct even though it has trailing padding, since 6 is the "packed" / binary.Size of that struct. Keep in mind that Go and C structures don't always have to correspond 1:1, for example Cilium doesn't use bpf2go and uses hand written Go types + a linter instead.

Is there something specific you're worried about with this "optimization"? Or is this because you think we shouldn't have the layout code at all?

> Trying to wrap my head around it, does it mean reading into correctly sized slice from bpf syscall directly? Will it work for updates as well?

That's what I would like, yes. The insight is that the layout of a per-CPU map for syscalls is:

```
[nCPU]struct{ [valueSize]byte; [align(valueSize, 8)-valueSize]byte }
```

`align(valueSize, 8)` is just `stride` in the layout code.

If you store something like `uint64` in the per CPU map there is no padding and the syscall could write into Go memory directly. This is really useful for storing counters, so having them be more efficient is always nice. OTOH we've been living with the reflect based decoding for a long time, so maybe there isn't such a pressing need.

Reply:
I tried your idea of using `binary.Size` instead of the custom logic in `layout.go`. It doesn't work because `binary.Size` doesn't reject unexported fields like in `struct{ a uint32 }`. We can't get rid of the custom reflect traversal unfortunately.